### PR TITLE
Serialization

### PR DESCRIPTION
**Pull request by blazej-smorawski**
Please review it thoroughly as it's quite an important piece of code.

### CODE REVIEW

File: usr/code.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+x = lambda y: y**2
```
Comment (filip-szweda) [inline usr/code.py]: `code.py` should stay empty on remote repository, this code serves only a purpouse for local testing.

File: src/point.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    # xxxXXXxxx WAAAAAAAAAARRRRRNINGGGGGGGG xxxXXXxxx
+    # Position is passed to properties as string,
+    # so frontends MUST parse the string on their own.
+    # It's not another JSON dictionary for reasons.
+    # That is - C# casting magic :)
+    def __str__(self) -> str:
+        return f"({self.x},{self.y})"
+
+    def __add__(self, other):
```
Comment (filip-szweda) [inline src/point.py]: Missing type hinting.

File: src/agent.py [hunk 3]
```diff
@@ -1,17 +1,10 @@
...
-        self.position = position
-
-    def move(self, direction: Direction) -> None:
-        print(f"MOVE {str(direction)}")
-        self.position = tuple(map(operator.add, self.position, direction.value))
-
-    def __str__(self) -> str:
-        return f"AGENT ({self.position[0]}, {self.position[1]})"
+class Agent(Object):
+    def __init__(self, position: Point = Point(0,0)):
```
Comment (filip-szweda) [inline src/agent.py]: `None` return type.

File: src/actions.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from .action import Action
+from .enums import Direction
+from .agent import Agent
+
+
+class Move(Action):
+    def __init__(self, agent: Agent, direction: Direction):
```
Comment (filip-szweda) [inline src/actions.py]: `None` return type.

File: main.py [hunk 5]
```diff
@@ -8,9 +8,9 @@
 
 
 parser = argparse.ArgumentParser(
-    prog='LNX-Runtime',
```
Comment (filip-szweda) [inline main.py]: Unnecessary change if we're going to use autoformatting.

File: src/point.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        self.x += other.x
+        self.y += other.y
+        return self
+
+    def dist_to(self, point) -> float:
```
Comment (filip-szweda) [inline src/point.py]: Missing type hinting.

File: src/runtime.py [hunk 7]
```diff
@@ -12,7 +14,7 @@ def run(self) -> None:
         agent = Agent()
         agent_locals = {}
         agent_builtins = {
-            'move': agent.move,
+            'move': lambda dir : Move(agent, dir).execute(),
```
Comment (filip-szweda) [inline src/runtime.py]: Name `dir` shadows builtin `dir` name and we've used `direction` name before. Let's stay consistent and use it here also.

File: src/action.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+from .object import Object
```
Comment (filip-szweda) [inline src/action.py]: Do you think we could give `./src` directory more structure? It is becoming difficult to read through.

File: src/serializable.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+# Properties is an empty class
+# used by objects inheriting from Serialzable
+# to store variables that will be serialized
+class Properties:
+    def __init__(self) -> None:
+        pass
+
+
+class Serializable():
+    def __init__(self, base) -> None:
```
Comment (filip-szweda) [inline src/serializable.py]: Missing type hinting.

Comment (filip-szweda) [inline src/serializable.py]: What is `base`? It doesn't tell me much.

File: src/actions.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+from .action import Action
```
Comment (filip-szweda) [inline src/actions.py]: A bit confusing file name, are we going to put all classes inheriting from `Action` here?

File: src/point.py (hunk 2 above)
Comment (blazej-smorawski) [reply src/point.py]: I think you cannot specify type `Point` inside `Point` class definition, but I might be wrong.

File: src/action.py (hunk 8 above)
Comment (blazej-smorawski) [reply src/action.py]: We should

**Comment by filip-szweda** [comment]
Consider writing tests and/or modifying existing ones.

**Comment by blazej-smorawski** [comment]
Sure thing, but I think it's not important at the stage of this projects development. And accidentally I pushed changes onto master, so I'm closing this one xD
